File: backend/app/routes/courses.py
from flask import Blueprint, jsonify

from app.models.course import (
    get_all_courses,
    get_course_by_id,
    get_popular_courses,
    serialize_course,
)
from app.models.teacher import get_teacher_by_id, serialize_teacher

courses_bp = Blueprint('courses', __name__, url_prefix='/api/courses')


# -------------------------------------------------------------------
# GET /api/courses
# -------------------------------------------------------------------
@courses_bp.route('', methods=['GET'])
def list_courses():
    courses = get_all_courses()
    return jsonify({
        "success": True,
        "data": [serialize_course(c) for c in courses],
    }), 200


# -------------------------------------------------------------------
# GET /api/courses/popular
# -------------------------------------------------------------------
@courses_bp.route('/popular', methods=['GET'])
def popular_courses():
    courses = get_popular_courses()
    return jsonify({
        "success": True,
        "data": [serialize_course(c) for c in courses],
    }), 200


# -------------------------------------------------------------------
# GET /api/courses/<course_id>
# -------------------------------------------------------------------
@courses_bp.route('/<course_id>', methods=['GET'])
def single_course(course_id):
    course = get_course_by_id(course_id)
    if not course:
        return jsonify({"success": False, "message": "Course not found"}), 404

    course_data = serialize_course(course)

    # Attach teacher info if available
    if course.get('teacher_id'):
        teacher = get_teacher_by_id(str(course['teacher_id']))
        course_data['teacher'] = serialize_teacher(teacher)

    return jsonify({
        "success": True,
        "data": course_data,
    }), 200

File: backend/routes/courses.js
import express from 'express';
import { getDB, ObjectId } from '../db.js';

const router = express.Router();

function serializeCourse(course) {
  if (!course) return null;
  return {
    id: course._id.toString(),
    title: course.title || '',
    slug: course.slug || '',
    description: course.description || '',
    short_description: course.short_description || '',
    icon: course.icon || 'BookOpen',
    image: course.image || '',
    level: course.level || 'Beginner',
    subjects: course.subjects || [],
    teacher_id: course.teacher_id ? course.teacher_id.toString() : '',
    is_popular: Boolean(course.is_popular),
    created_at: course.created_at ? new Date(course.created_at).toISOString() : null,
  };
}

function serializeTeacher(teacher) {
  if (!teacher) return null;
  return {
    id: teacher._id.toString(),
    name: teacher.name || teacher.full_name || '',
    title: teacher.title || '',
    bio: teacher.bio || '',
    experience_years: teacher.experience_years || 0,
    image: teacher.image || '',
    specializations: teacher.specializations || [],
  };
}

// GET /api/courses
router.get('/', async (req, res) => {
  try {
    const db = getDB();
    const courses = await db.collection('courses').find().toArray();
    return res.json({
      success: true,
      data: courses.map(serializeCourse),
    });
  } catch (err) {
    return res.status(500).json({ success: false, message: 'Server error' });
  }
});

// GET /api/courses/popular
router.get('/popular', async (req, res) => {
  try {
    const db = getDB();
    const courses = await db.collection('courses').find({ is_popular: true }).toArray();
    return res.json({
      success: true,
      data: courses.map(serializeCourse),
    });
  } catch (err) {
    return res.status(500).json({ success: false, message: 'Server error' });
  }
});

// GET /api/courses/:course_id
router.get('/:course_id', async (req, res) => {
  try {
    const db = getDB();
    let course;
    try {
      course = await db.collection('courses').findOne({ _id: new ObjectId(req.params.course_id) });
    } catch {
      course = await db.collection('courses').findOne({ slug: req.params.course_id });
    }

    if (!course) {
      return res.status(404).json({ success: false, message: 'Course not found' });
    }

    const courseData = serializeCourse(course);

    if (course.teacher_id) {
      try {
        const teacher = await db.collection('teachers').findOne({ _id: new ObjectId(course.teacher_id) })
                     || await db.collection('users').findOne({ _id: new ObjectId(course.teacher_id) });
        if (teacher) {
          courseData.teacher = serializeTeacher(teacher);
        }
      } catch {}
    }

    return res.json({ success: true, data: courseData });
  } catch (err) {
    return res.status(500).json({ success: false, message: 'Server error' });
  }
});

export default router;
